### DISCUSSION TITLE
[A-L] (19 febbraio 2024) Esercizio 1

### DISCUSSION THREAD

**Discussion by Frisbon** [sapienzastudentsnetwork/progettazione-di-algoritmi#83] [Esami A-L]
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/a4758a40-fdbc-4278-a35a-c1c0dbfe022c)

**Comment by Frisbon** [comment]
Non so fare le dimostrazioni ma ho dato degli esempi che possono aiutare (e spero mi guadagnino punti all'esame)

![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/aec7d3d0-623e-4bbf-9c5b-3c25c28acab3)
----------
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/36790c6b-28b8-4015-be66-23d826a92e07)

**Comment by FeddyLix17** [comment]
Scelto un certo grafo non diretto connesso e pesato G (ho numerato i vertici alfabeticamente per evitare confusione)

![photo_2024-06-17_17-37-46](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/66442720/b239fa31-d873-4fda-987a-cee284c381b1)

del quale è stato calcolato

- il minimo albero di copertura (Minimum Spanning Tree) $\Large T_1$ (nell'immagine che segue tramite l'algoritmo di [Prim](https://www.programiz.com/dsa/prim-algorithm))

![photo_2024-06-17_18-29-59](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/66442720/bf5af4f9-990d-4fee-b05b-1063d00daf2f)

- l'albero dei cammini minimi $\Large T_2$ dal nodo 1 (nodo A) agli altri nodi di G (nell'immagine che segue tramite l'algoritmo di [Dijkstra](https://www.programiz.com/dsa/dijkstra-algorithm))

![photo_2024-06-17_19-02-37](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/66442720/d4582d12-b28d-481d-9410-8c4d3cd731fa)

Considerando ora il grafo G' ottenuto da G incrementando il costo di ciascun arco di G di una stessa quantità (viene scelto 1 come valore)

![photo_2024-06-17_19-26-34](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/66442720/b8b31760-18a9-43f3-9df2-cc84331bddda)

applicando gli algoritmi sopra riportati si ottengono rispettivamente i seguenti Minimum Spanning Tree, e Shortest Path dal vertice A 

![photo_2024-06-17_22-31-43](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/66442720/e697b487-834a-4fae-b36b-f3b538029891)

![photo_2024-06-17_19-36-48](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/66442720/2ff02449-d6fe-4bfd-bce8-6cd84a255981)

risultando che

$$
    \Large T_1 = T_1^{'} \wedge T_2= T_2^{'} 
$$

Partendo sempre dal vertice A, $\Large \forall quantità \in \mathbb{R}$ positivo, le condizioni $\Large a)$ e $\Large b)$ risultano vere.

Tuttavia ci sono casi che non rendono sempre vere entrambe le affermazioni (come riportato nell'esempio di @Frisbon )

In conclusione

- a) è vera SEMPRE

- b) è vera SE E SOLO SE  dato il vertice di partenza dell'algoritmo Dijkstra $\Large u  \in V(G),\ \forall (u,\ v) \in E(G)$, dove $\Large v \in V(G)\ |\ u \neq v$ ogni cammino possibile per raggiungere quel particolare nodo $v$ abbia lo stesso numero di vertici (e di conseguenza di archi)

Nel mio caso, la condizione $\Large b)$ sarà sempre $\Large \underline{vera}$ qualora i nodi di partenza per l'algoritmo di Dijkstra siano A, C, F oppure H,  $\Large \underline{falsa}$  altrimenti, mentre nel caso di @Frisbon dovrebbero essere il nodo centrale alto/basso.

**Comment by cornflexxx** [comment]
$a)$ Sia $S(E(G))$ l'array con gli archi di $G$ ordinati in ordine crescente prima di essere aggiunti all'MST:

$$ S(E(G)) : = [ e_1,e_2,...,e_k] \qquad \text{con } w(e_1)\le w(e_2 )\le ...\le w( e_k )$$

Definita $w'(e_i) := w(e_i) + k$, quindi  $w(e_i)\le w(e_j) \Longrightarrow w'(e_i)\le w'(e_j)$. Quindi ogni elemento in $S(E(G'))$ manterrà la posizione che aveva in  $S(E(G))$, ne segue che ogni arco verrà aggiunto a $T_1'$ con lo stesso ordine che in $T_1$ e quindi  $E(T_1) = E(T_1')$.

$b)$ Nel secondo caso  l'affermazione è falsa, controesempio:
<img width="268" alt="Screenshot 2024-07-14 alle 16 44 44" src="https://github.com/user-attachments/assets/cba40207-2ba8-4be7-bee6-e6992876fa8a">
$T_2 := \\{ (u,v),(v,s)\\}$
Cambiamo ora $w$ con $w' := w(e)+1$
$T_2' :=\\{ (u,v),(u,s)\\}$

In generale supponiamo che $P(x,y)$ sia un cammino minimo da $x\rightarrow y$ con peso  $w(P) = k$, e sia $w'(e) := w(e) + t$. Allora se applichiamo $w'$ a $P$ otterremo : 

$$ w'(P) = w(P) + t|E(P)|$$

Prendiamo ora un cammino $Q$ con $w$-peso $w(Q)= k + c$ e con $w'(Q) = k + c + t|E(Q)|$. Il cammino $Q$ misurato con $w'$ può essere un nuovo cammino minimo se e solo se $t|E(Q)| + c < t|E(P)|$, ovvero nel caso in cui il numero di archi di $P$ non lo renda più un cammino minimo valido.

**Comment by daromacarlo** [comment]
![immagine](https://github.com/user-attachments/assets/cd28c6a7-a350-47a7-aedd-f2b2d16af812)
